### PR TITLE
better fluid spreading

### PR DESCRIPTION
**Pull request by Slava0135**
## About the PR

Makes fluid spreading use raycasting from source puddle to possible new puddle location.

**Media**

Peek 2023-03-08 20-10

**Changelog**

:cl:
- fix: fluid can now spread through directional windows (through not occupied sides).

### CODE REVIEW

**Comment by Snowyypanda** [comment]
can't you just rip off ss13 tg station fluids animations? they're waay better than what there currently is in ss14

**Comment by Slava0135** [comment]
> can't you just rip off ss13 tg station fluids animations? they're waay better than what there currently is in ss14

Point was spreading itself

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/Fluids/EntitySystems/FluidSpreaderSystem.cs [hunk 1]
```diff
@@ -100,49 +101,55 @@ public override void Update(float frameTime)
...
+            newPuddleComp = null;
             return false;
         }
 
-        var puddleCurrentVolume = _puddleSystem.CurrentVolume(srcUid, srcPuddle);
+        // check if puddle can spread there at all
+        var dst = dstPos.ToMapPos(EntityManager, _transform);
+        var src = Transform(srcUid).Coordinates.ToMapPos(EntityManager, _transform);
+        var dir = src - dst;
+        var ray = new CollisionRay(dst, dir.Normalized, (int) CollisionGroup.MobMask);
```
[the remark this answers is not in the document]
Comment (Slava0135) [reply Content.Server/Fluids/EntitySystems/FluidSpreaderSystem.cs:123]: i took it from old code, no idea what it should be

File: Content.Server/Fluids/EntitySystems/FluidSpreaderSystem.cs [hunk 2]
```diff
@@ -100,49 +101,57 @@ public override void Update(float frameTime)
...
             return false;
         }
 
-        var puddleCurrentVolume = _puddleSystem.CurrentVolume(srcUid, srcPuddle);
+        // check if puddle can spread there at all
+        var dstMap = dstPos.ToMap(EntityManager, _transform);
+        var dst = dstMap.Position;
+        var src = Transform(srcUid).MapPosition.Position;
+        var dir = src - dst;
+        var ray = new CollisionRay(dst, dir.Normalized, (int) CollisionGroup.TabletopMachineMask);
```
Comment (ElectroJr) [inline Content.Server/Fluids/EntitySystems/FluidSpreaderSystem.cs:125]: Why tabletop mask? Shouldn't it be something like `LowImpassable | Impassable`?

Comment (Slava0135) [reply Content.Server/Fluids/EntitySystems/FluidSpreaderSystem.cs:125]: @metalgearsloth wanted to change it

**Comment by Slava0135** [comment]
Peek 2023-03-10 20-50

File: Content.Server/Fluids/EntitySystems/FluidSpreaderSystem.cs (hunk 2 above)
Comment (metalgearsloth) [reply Content.Server/Fluids/EntitySystems/FluidSpreaderSystem.cs:125]: Try not to use the composite masks unless there's a good reason, just go with a flag of individual layers.

**Pull request merged**